express-notes-api/index.js

```diff
@@ -0,0 +1,158 @@
+
+var express = require('express');
+var app = express();
+var filedata = require('./data.json');
+var fs = require('fs');
+const { restart } = require('nodemon');
+
+app.listen(3000, () => {
+  // eslint-disable-next-line no-console
+  console.log('Express server listening on port 3000');
+});
+
+/*** errors */
+
+function error400PositiveInteger (res) {
+  res.status(400);
+  res.json({ "error": "id must be a positive integer" });
+}
+
+function error404CantFindID(res, id) {
+  res.status(404);
+  res.json({ "error": `cannot find note with id ${id}` });
+}
+
+function error404ContentRequiredField (res) {
+  res.status(400);
+  res.json({ "error": "content is a required field" });
+}
+
+function unexpectedError500 (res) {
+  res.status(500);
+  res.json({ "error": "An unexpected error occurred." });
+}
+
+
+/*** Clients can GET a list of notes */
+
+app.get('/api/notes', function (req, res) {
+  var arrayNotes = [];
+  for (const value in filedata.notes){
+    arrayNotes.push(filedata.notes[value]);
+  }
+  res.json(arrayNotes);
+});
+
+/** Clients can GET a single note by id */
+
+app.get('/api/notes/:id', function (req, res) {
+  var isThereANote = false;
+  var id = parseInt(req.params.id);
+
+  if (Number.isNaN(id)) {
+    error400PositiveInteger(res);
+  } else {
+    for (const value in filedata.notes) {
+      if (filedata.notes[value].id === id) {
+        res.json(filedata.notes[value]);
+        isThereANote = true;
+        break;
+      }
+    }
+    if (isThereANote === false) {
+      error404CantFindID(res, id);
+    }
+  }
+});
+
+/***  Clients can POST a new note */
+
+app.use(express.json());
+app.post('/api/notes', function (req, res) {
+
+  if (req.body.content === undefined) {
+    error404ContentRequiredField(res);
+  } else {
+    var id = filedata.nextId;
+    filedata.notes[id] = req.body;
+    filedata.notes[id].id = id;
+    fs.writeFile('data.json', JSON.stringify(filedata, null, 2), 'utf8', (err) => {
+      if (err) {
+        unexpectedError500(res);
+      } else {
+        res.status(201);
+        res.json(filedata.notes[id]);
+      }
+    });
+    filedata.nextId = filedata.nextId + 1;
+  }
+});
+
+/*** Clients can DELETE a note by id */
+
+app.delete('/api/notes', function (req, res) {
+  error400PositiveInteger(res);
+});
```

Comment:
This is more of a bad URL error than a invalid `id` error, since no `id` is being provided, so the default `404 Not Found` would be a more appropriate response.